src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineCanvas/TemplatePipelineCanvas.tsx

```diff
@@ -0,0 +1,110 @@
+import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
+import { useStrings } from 'framework/strings'
+import { useValidationErrors } from '@pipeline/components/PipelineStudio/PiplineHooks/useValidationErrors'
+import { CanvasButtons } from '@pipeline/components/CanvasButtons/CanvasButtons'
+import { getScopeBasedProjectPathParams, getScopeFromValue } from '@common/components/EntityReference/EntityReference'
+import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
+import { useGetYamlWithTemplateRefsResolved } from 'services/template-ng'
+import { useMutateAsGet } from '@common/hooks'
+import { yamlStringify } from '@common/utils/YamlHelperMethods'
+import css from './TemplatePipelineCanvas.module.scss'
+
+export function TemplatePipelineCanvas(): React.ReactElement {
+  const {
+    state: { pipeline, templateTypes, gitDetails },
+    stagesMap
+  } = usePipelineContext()
+  const canvasRef = React.useRef<HTMLDivElement | null>(null)
+  const { getString } = useStrings()
+  const { errorMap } = useValidationErrors()
+  const [resolvedPipeline, setResolvedPipeline] = React.useState<PipelineInfoConfig>()
+  const scope = getScopeFromValue(defaultTo(pipeline.template?.templateRef, ''))
+  const queryParams = useParams<ProjectPathProps>()
+  const dummyPipeline = useRef(pipeline)
+
+  const model = React.useMemo(() => new StageBuilderModel(), [])
+  const engine = React.useMemo(() => createEngine({}), [])
+  engine.setModel(model)
+
+  const addUpdateGraph = () => {
+    model.addUpdateGraph({
+      data: resolvedPipeline || { name: '', identifier: '' },
+      listeners: {
+        nodeListeners: {},
+        linkListeners: {}
+      },
+      zoomLevel: model.getZoomLevel(),
+      stagesMap,
+      getString,
+      isReadonly: true,
+      splitPaneSize: canvasRef.current?.offsetHeight,
+      parentPath: 'pipeline.stages',
+      errorMap,
+      templateTypes
+    })
+  }
+
+  addUpdateGraph()
+
+  useStageBuilderCanvasState(engine, [])
+
+  const {
+    data: pipelineResponse,
+    error: pipelineError,
+    refetch: refetchPipeline,
+    loading: pipelineLoading
+  } = useMutateAsGet(useGetYamlWithTemplateRefsResolved, {
+    queryParams: {
+      ...getScopeBasedProjectPathParams(queryParams, scope),
+      pipelineIdentifier: pipeline.identifier,
+      repoIdentifier: gitDetails.repoIdentifier,
+      branch: gitDetails.branch,
+      getDefaultFromOtherRepo: true
+    },
+    body: {
+      originalEntityYaml: yamlStringify(dummyPipeline.current)
+    }
+  })
+
+  React.useEffect(() => {
+    if (pipelineResponse?.data?.mergedPipelineYaml) {
+      setResolvedPipeline(parse(pipelineResponse?.data?.mergedPipelineYaml))
+    }
+  }, [pipelineResponse?.data?.mergedPipelineYaml])
+
+  React.useEffect(() => {
+    dummyPipeline.current = pipeline
+  }, [pipeline.template?.templateRef, pipeline.template?.versionLabel])
```

Comment:
Codacy has a fix for the issue: [React Hook React.useEffect has a missing dependency: 'pipeline'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [pipeline, pipeline.template.templateRef, pipeline.template.versionLabel])
```